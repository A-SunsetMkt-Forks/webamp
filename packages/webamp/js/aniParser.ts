import { RIFFFile } from "riff-file";
import { unpackArray, unpackString } from "byte-data";

type Chunk = {
  format: string;
  chunkId: string;
  chunkData: {
    start: number;
    end: number;
  };
  subChunks: Chunk[];
};

// https://www.informit.com/articles/article.aspx?p=1189080&seqNum=3
export type AniMetadata = {
  cbSize: number; // Data structure size (in bytes)
  nFrames: number; // Number of images (also known as frames) stored in the file
  nSteps: number; // Number of frames to be displayed before the animation repeats
  iWidth: number; // Width of frame (in pixels)
  iHeight: number; // Height of frame (in pixels)
  iBitCount: number; // Number of bits per pixel
  nPlanes: number; // Number of color planes
  iDispRate: number; // Default frame display rate (measured in 1/60th-of-a-second units)
  bfAttributes: number; // ANI attribute bit flags
};

type AniInfo = { title: string | null; artist: string | null };

export type ParsedAni = {
  rate: number[] | null;
  seq: number[] | null;
  images: Uint8Array[];
  metadata: AniMetadata;
  artist: string | null;
  title: string | null;
};

const DWORD = { bits: 32, be: false, signed: false, fp: false };

export function parseAni(arr: Uint8Array): ParsedAni {
  const riff = new RIFFFile();

  riff.setSignature(arr);

  const signature = riff.signature as Chunk;
  if (signature.format !== "ACON") {
    throw new Error(`Expected fromat "ACON", got "${signature.format}"`);
  }

  let metadata: null | AniMetadata = null;
  let rate: number[] | null = null;
  let seq: number[] | null = null;
  let images: Uint8Array[] | null = null;
  let info: AniInfo = { artist: null, title: null };

  signature.subChunks.forEach(({ chunkId, chunkData, subChunks, format }) => {
    switch (trimNullTerminated(chunkId)) {
      case "anih":
        metadata = parseMetadata(arr, chunkData.start, chunkData.end);
        break;
      case "rate":
        rate = unpackArray(arr, DWORD, chunkData.start, chunkData.end);
        break;
      case "seq":
        seq = unpackArray(arr, DWORD, chunkData.start, chunkData.end);
        break;
      case "LIST": // TODO: assert(i === subChunks.length)
        switch (format) {
          case "INFO":
            info = parseInfo(arr, subChunks);
            break;
          case "fram":
            images = subChunks.map((c) => {
              if (c.chunkId !== "icon") {
                throw new Error(`Unexpected chunk type in fram: ${chunkId}`);
              }
              return arr.slice(c.chunkData.start, c.chunkData.end);
            });
        }
        break;
      default:
      // TODO: We could assert that this never happens
    }
  });

  if (metadata == null) {
    throw new Error("Did not find anih");
  }

  if (images == null) {
    throw new Error("Did not find LIST");
  }

  return { ...info, images, rate, seq, metadata };
}

function parseInfo(arr: Uint8Array, chunks: Chunk[]): AniInfo {
  const info: AniInfo = { title: null, artist: null };
  chunks.forEach((chunk) => {
    switch (chunk.chunkId) {
      case "INAM":
        info.title = trimNullTerminated(
          unpackString(arr, chunk.chunkData.start, chunk.chunkData.end)
        );
        break;
      case "IART":
        info.artist = trimNullTerminated(
          unpackString(arr, chunk.chunkData.start, chunk.chunkData.end)
        );
        break;
      default:
      // Unexpected subchunk
    }
  });
  return info;
}

function parseMetadata(
  arr: Uint8Array,
  start: number,
  end: number
): AniMetadata {
  // TODO: We could assert that we have 9 items here.
  const words = unpackArray(arr, DWORD, start, end);
  return {
    cbSize: words[0],
    nFrames: words[1],
    nSteps: words[2],
    iWidth: words[3],
    iHeight: words[4],
    iBitCount: words[5],
    nPlanes: words[6],
    iDispRate: words[7],
    bfAttributes: words[8],
  };
}

// I suspect that RIFF points to byte ranges, but that includes the byte(s?)
// used for null termination.
function trimNullTerminated(str: string): string {
  return str.trim();
}
