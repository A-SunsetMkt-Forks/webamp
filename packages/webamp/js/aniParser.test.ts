import fs from "fs";
import path from "path";
import { parseAni } from "./aniParser";

// Parse a `.ani` in our fixture directory and trim down the image data for use
// in snapshot tests.
function parsePath(filePath: string) {
  const buffer = fs.readFileSync(
    path.join(__dirname, "./__tests__/fixtures/ani/", filePath)
  );

  const ani = parseAni(buffer);
  const imgString = ani.images.map((image) => image.slice(0, 60).join(""));
  // @ts-ignore
  ani.images = `${imgString}...`;
  return ani;
}

// https://skins.webamp.org/skin/6e30f9e9b8f5719469809785ae5e4a1f/Super_Mario_Amp_2.wsz/
describe("Super_Mario_Amp_2.wsz", () => {
  test("eqslid.cur", async () => {
    expect(parsePath("Super_Mario_Amp_2/eqslid.cur")).toMatchSnapshot();
  });
  test("close.cur", async () => {
    expect(parsePath("Super_Mario_Amp_2/close.cur")).toMatchSnapshot();
  });
});

// https://skins.webamp.org/skin/4308a2fc648033bf5fe7c4d56a5c8823/Green-Dimension-V2.wsz/
describe("Green Dimension v2.wsz", () => {
  test("eqslid.cur", async () => {
    expect(parsePath("Green Dimension v2/eqslid.cur")).toMatchSnapshot();
  });
});
