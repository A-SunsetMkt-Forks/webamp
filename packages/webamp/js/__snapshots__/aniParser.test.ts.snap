// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Green Dimension v2.wsz eqslid.cur 1`] = `
Object {
  "images": "00201032320000001681600220004000032000640001032000000160000000000000000,00201032320000001681600220004000032000640001032000000160000000000000000...",
  "metadata": Object {
    "bfAttributes": 1,
    "cbSize": 36,
    "iBitCount": 0,
    "iDispRate": 10,
    "iHeight": 0,
    "iWidth": 0,
    "nFrames": 2,
    "nPlanes": 0,
    "nSteps": 2,
  },
  "rate": Array [
    8,
    8,
  ],
  "seq": null,
}
`;

exports[`Super_Mario_Amp_2.wsz close.cur 1`] = `
Object {
  "images": "0020103232000000168800220004000032000640001080000012840000000000000000,0020103232000000168800220004000032000640001080000012840000000000000000,0020103232000000168800220004000032000640001080000012840000000000000000,0020103232000000168800220004000032000640001080000012840000000000000000,0020103232000000168800220004000032000640001080000012840000000000000000,0020103232000000168800220004000032000640001080000012840000000000000000,0020103232000000168800220004000032000640001080000012840000000000000000,0020103232000000168800220004000032000640001080000012840000000000000000,0020103232000000168800220004000032000640001080000012840000000000000000,0020103232000000168800220004000032000640001080000012840000000000000000...",
  "metadata": Object {
    "bfAttributes": 3,
    "cbSize": 36,
    "iBitCount": 4,
    "iDispRate": 10,
    "iHeight": 0,
    "iWidth": 0,
    "nFrames": 10,
    "nPlanes": 1,
    "nSteps": 17,
  },
  "rate": Array [
    10,
    10,
    10,
    10,
    7,
    6,
    5,
    4,
    3,
    2,
    1,
    2,
    2,
    2,
    2,
    2,
    60,
  ],
  "seq": Array [
    0,
    1,
    2,
    3,
    4,
    5,
    4,
    5,
    4,
    5,
    6,
    7,
    8,
    9,
    8,
    9,
    8,
  ],
}
`;

exports[`Super_Mario_Amp_2.wsz eqslid.cur 1`] = `
Object {
  "images": "0020103232000000168120022000400003200064000102400000128120000000000000000,0020103232000000168120022000400003200064000102400000128120000000000000000,0020103232000000168120022000400003200064000102400000128120000000000000000,0020103232000000168120022000400003200064000102400000128120000000000000000,0020103232000000168120022000400003200064000102400000128120000000000000000,0020103232000000168120022000400003200064000102400000128120000000000000000,0020103232000000168120022000400003200064000102400000128120000000000000000,0020103232000000168120022000400003200064000102400000128120000000000000000...",
  "metadata": Object {
    "bfAttributes": 1,
    "cbSize": 36,
    "iBitCount": 4,
    "iDispRate": 10,
    "iHeight": 0,
    "iWidth": 0,
    "nFrames": 8,
    "nPlanes": 1,
    "nSteps": 8,
  },
  "rate": null,
  "seq": null,
}
`;
